### DISCUSSION TITLE
Creating hierarchical processing modules in NWBfiles

### DISCUSSION THREAD

**Discussion by mudd54** [dandi/helpdesk#153] [Q&A]
Hello, 

I’m trying to create some nwbs. I would like this hierarchy of processing modules: 

```
/ nwbfile
└── processing
    ├── probe_1234
    │   ├── karlsonevents
    │   │   └── timeintervals
    │   ├── gammaevents
    │   │   └── timeintervals
    │   └── other events modules etc.
    ├── probe_5678
    │   └── ...
    └── ...
```

Where probe_1234 is a high-level processing module containing the low-level processing modules karlssonevent and gammaevents (and other event modules). karlsonevent and gammaevents contain timeinterval objects. 

However, whenever I try to set up the processing module hierarchy, I end up with the events processing modules at the same level as the probe processing modules, and also duplicated inside the probe processing modules. like so: 

```
/ nwbfile
└── processing
    ├── probe_1234
    │   ├── karlsonevents
    │   │   └── timeintervals
    │   ├── gammaevents
    │   │   └── timeintervals
    │   └── other events modules etc.
    ├── karlsonevents
    │   └── timeintervals
    ├── gammaevents
    │   └── timeintervals
    └── probe_5678
        └── ...
```
![image](https://github.com/dandi/helpdesk/assets/168661874/c70151a8-ff11-48b7-88e4-26a94d621c0f)

I assume this is because processing modules must be created through nwbfile.create_processing_module() and can’t be instantiated directly. 

**My question is: is there a way to set up the nwbfile in my desired processing module hierarchy?** Is this the correct way of using processing modules in the nwb? This is my first time working with them. Thank you.

**Comment by CodyCBakerPhD** [comment]
attn @rly

**Comment by kabilar** [comment]
Hi @bendichter @h-mayorquin, could you please look into @mudd54's questions?  Thanks.

**Comment by h-mayorquin** [comment]
Can you provide the code that you are using to create the hierarchy @mudd54 ?

**Comment by bendichter** [comment]
Hi @mudd54. There is not currently a way to store a `ProcessingModule` within another `ProcessingModule`. I would recommend either creating multiple processing modules on the same level or using the names of the neurodata objects to communicate their associations.

**Reply by bendichter** [reply]
If you think NWB should allow this, I encourage you to raise an issue on the NWB schema repo [here](https://github.com/NeurodataWithoutBorders/nwb-schema/issues)
